cassandra/src/main/java/se/tre/freki/storage/cassandra/search/CassandraSearchPlugin.java

```diff
@@ -90,25 +95,117 @@ public void close() throws IOException {
   @Override
   public ListenableFuture<Void> deleteLabelMeta(final LabelId id,
                                                 final LabelType type) {
-    return Futures.immediateFuture(null);
+    return immediateFuture(null);
   }
 
   @Nonnull
   @Override
   public ListenableFuture<Void> indexAnnotation(final Annotation note) {
-    return Futures.immediateFuture(null);
+    return immediateFuture(null);
   }
 
   @Nonnull
   @Override
   public ListenableFuture<Void> deleteAnnotation(final Annotation note) {
-    return Futures.immediateFuture(null);
+    return immediateFuture(null);
   }
 
   @Nonnull
   @Override
   public ListenableFuture<Iterable<LabelMeta>> findLabels(final String query) {
-    return Futures.<Iterable<LabelMeta>>immediateFuture(ImmutableSet.<LabelMeta>of());
+    NGramGenerator ngrams = new NGramGenerator(query);
+    List<ResultSetFuture> resultSets = fetchLabelIds(ngrams);
+
+    return transform(
+        successfulAsList(resultSets),
+        new AsyncFunction<List<ResultSet>, Iterable<LabelMeta>>() {
+          @Override
+          public ListenableFuture<Iterable<LabelMeta>> apply(final List<ResultSet> resultSets) {
+            return transform(fetchLabelMetas(resultSets),
+                new Function<Iterable<LabelMeta>, Iterable<LabelMeta>>() {
+                  @Override
+                  public Iterable<LabelMeta> apply(final Iterable<LabelMeta> input) {
+                    return input;
+                  }
+                });
+          }
+        });
+  }
+
+  /**
+   * Takes a List of ResultSets and for each labelid in the resulset
+   * get the corresponding LabelMeta.
+   *
+   * @param resultSets A List of ResultSets
+   *
+   * @return A ListenableFuture with a list of LabelMetas.
+   */
+  protected ListenableFuture<? extends Iterable<LabelMeta>> fetchLabelMetas(
+      final List<ResultSet> resultSets) {
+    List<ListenableFuture<LabelMeta>> metas = new ArrayList<>();
+
+    for (ResultSet resultSet : resultSets) {
+      for (Row row : resultSet) {
+        CassandraLabelId id = CassandraLabelId.fromLong(row.getLong(1));
```

Comment:
Use the name of the field instead of the index of it on the call to `#getLong`. It makes it clearer which field we get and it prevents potentially hard to find bugs if we ever switch the order of any columns